File: native/zenohex_nif/src/keyexpr.rs
#[rustler::nif]
fn keyexpr_autocanonize(key_expr: String) -> rustler::NifResult<(rustler::Atom, String)> {
    let mut key_expr = key_expr;
    match zenoh::key_expr::keyexpr::autocanonize(&mut key_expr) {
        Ok(key_expr) => Ok((rustler::types::atom::ok(), key_expr.to_string())),
        Err(error) => Err(rustler::Error::Term(crate::zenoh_error!(error))),
    }
}

#[rustler::nif(name = "keyexpr_valid?")]
fn keyexpr_valid(key_expr: &str) -> rustler::NifResult<bool> {
    match zenoh::key_expr::keyexpr::new(key_expr) {
        Ok(_) => Ok(true),
        Err(error) => {
            log::error!("{}", error);
            Ok(false)
        }
    }
}

#[rustler::nif(name = "keyexpr_intersects?")]
fn keyexpr_intersects(key_expr1: &str, key_expr2: &str) -> rustler::NifResult<bool> {
    let key_expr1 = new(key_expr1)?;
    let key_expr2 = new(key_expr2)?;

    Ok(key_expr1.intersects(key_expr2))
}

#[rustler::nif(name = "keyexpr_includes?")]
fn keyexpr_includes(key_expr1: &str, key_expr2: &str) -> rustler::NifResult<bool> {
    let key_expr1 = new(key_expr1)?;
    let key_expr2 = new(key_expr2)?;

    Ok(key_expr1.includes(key_expr2))
}

#[rustler::nif]
fn keyexpr_join(key_expr1: &str, key_expr2: &str) -> rustler::NifResult<(rustler::Atom, String)> {
    let ke1 = zenoh::key_expr::keyexpr::new(key_expr1)
        .map_err(|error| rustler::Error::Term(crate::zenoh_error!(error)))?;
    ke1.join(key_expr2)
        .map(|ke| (rustler::types::atom::ok(), ke.to_string()))
        .map_err(|error| rustler::Error::Term(crate::zenoh_error!(error)))
}

fn new(key_expr: &str) -> rustler::NifResult<&zenoh::key_expr::keyexpr> {
    match zenoh::key_expr::keyexpr::new(key_expr) {
        Ok(key_expr) => Ok(key_expr),
        Err(error) => Err(rustler::Error::RaiseTerm(Box::new(
            crate::helper::exception::ArgumentError {
                message: error.to_string(),
            },
        ))),
    }
}

File: lib/zenohex/key_expr.ex
defmodule Zenohex.KeyExpr do
  @moduledoc """
  Provides utility functions for working with Zenoh key expressions (keyexpr).

  Key expressions in Zenoh define paths or patterns used to match resources.
  These utilities help manage their correctness and relationships.
  """

  @doc """
  Canonizes the `key_expr`.

  ## Examples

      iex> Zenohex.KeyExpr.canonize("key/expr/**/*")
      {:ok, "key/expr/*/**"}
  """
  @spec canonize(String.t()) :: {:ok, String.t()} | {:error, String.t()}
  defdelegate canonize(key_expr),
    to: Zenohex.Nif,
    as: :keyexpr_autocanonize

  @doc """
  Validate the `key_expr`.
  """
  @spec valid?(String.t()) :: boolean()
  defdelegate valid?(key_expr),
    to: Zenohex.Nif,
    as: :keyexpr_valid?

  @doc """
  Returns true if the keyexprs intersect.

  There exists at least one key which is contained in both of the sets defined by `keyr_expr1` and `key_expr2`.
  """
  @spec intersects?(String.t(), String.t()) :: boolean()
  defdelegate intersects?(key_expr1, key_expr2),
    to: Zenohex.Nif,
    as: :keyexpr_intersects?

  @doc """
  Returns true if `key_expr1` includes `key_expr2`.

  The set defined by `key_expr1` contains every key belonging to the set defined by `key_expr2`.
  """
  @spec includes?(String.t(), String.t()) :: boolean()
  defdelegate includes?(key_expr1, key_expr2),
    to: Zenohex.Nif,
    as: :keyexpr_includes?

  @doc """
  Joins two key expressions, inserting a `/` between them.

  Returns the resulting key expression string, canonized.

  ## Examples

      iex> Zenohex.KeyExpr.join("key/expr", "sub")
      {:ok, "key/expr/sub"}
  """
  @spec join(String.t(), String.t()) :: {:ok, String.t()} | {:error, String.t()}
  defdelegate join(key_expr1, key_expr2),
    to: Zenohex.Nif,
    as: :keyexpr_join
end
